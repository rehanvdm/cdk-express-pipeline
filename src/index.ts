export * from './cdk-express-pipeline';
export * from './cdk-express-pipeline-legacy';
export * from './shared';
export * from './express-stack';
export * from './express-stage';
export * from './express-wave';


