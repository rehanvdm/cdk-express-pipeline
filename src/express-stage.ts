import { ExpressStack } from './express-stack';
import { ExpressWave } from './express-wave';

export interface IExpressStage {
  /**
   * The stage identifier
   */
  id: string;

  /**
   * The wave that the stage belongs to
   */
  wave: ExpressWave;

  /**
   * The stacks in the stage
   */
  stacks: ExpressStack[];

}


/**
 * A CDK Express Pipeline Stage that belongs to an ExpressWave
 */
export class ExpressStage implements IExpressStage {
  public id: string;
  public wave: ExpressWave;
  public stacks: ExpressStack[] = [];

  /**
   * Constructs a new instance of the ExpressStage class
   * @param id The stage identifier
   * @param wave The wave that the stage belongs to
   * @param stacks The ExpressStacks in the stage
   */
  constructor(id: string, wave: ExpressWave, stacks: ExpressStack[] = []) {
    this.id = id;
    this.wave = wave;
    this.wave.stages.push(this);
    this.stacks = stacks;
    if (this.id.includes(this.wave.separator)) {
      throw new Error(`ExpressStage '${id}' cannot contain a '${this.wave.separator} ' (separator)`);
    }
  }
}