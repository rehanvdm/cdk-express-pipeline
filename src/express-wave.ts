import { CDK_EXPRESS_PIPELINE_DEFAULT_SEPARATOR } from './cdk-express-pipeline';
import { ExpressStage } from './express-stage';

export interface IExpressWave {
  /**
   * The wave identifier
   */
  id: string;

  /**
   * Separator between the wave, stage and stack ids that are concatenated to form the final stack id
   */
  separator: string;

  /**
   * The ExpressStages in the wave
   */
  stages: ExpressStage[];

  /**
   * Add an ExpressStage to the wave
   * @param id The ExpressStage identifier
   */
  addStage(id: string): ExpressStage;
}

export interface ExpressWaveProps {
  readonly id: string;

  /**
   * Separator between the wave, stage and stack ids that are concatenated to form the stack id
   * @default `_`
   */
  readonly separator?: string;
}

/**
 * A CDK Express Pipeline Wave that contains ExpressStages
 */
export class ExpressWave implements IExpressWave {
  public id: string;
  public separator: string;
  public stages: ExpressStage[] = [];

  /**
   * Constructs a new instance of the ExpressWave class
   * @param id
   * @param separator Separator between the wave, stage and stack ids that are concatenated to form the stack id. Default: '_'.
   */
  constructor(id: string, separator: string = CDK_EXPRESS_PIPELINE_DEFAULT_SEPARATOR) {
    this.id = id;
    this.separator = separator;
    if (this.id.includes(this.separator)) {
      throw new Error(`ExpressWave '${this.id}' cannot contain a '${this.separator}' (separator)`);
    }
  }

  public addStage(id: string): ExpressStage {
    return new ExpressStage(id, this);
  }
}