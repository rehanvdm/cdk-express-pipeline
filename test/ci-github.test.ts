import * as fsSync from 'fs';
import * as fs from 'fs/promises';
import * as path from 'path';
import { App } from 'aws-cdk-lib';
import { stringify } from 'yaml';
import { CdkExpressPipeline } from '../src/cdk-express-pipeline';
import { GitHubWorkflowConfig, createGitHubWorkflows, BuildWorkflowPathConfig } from '../src/ci-github';
import { ExpressStack } from '../src/express-stack';

function getAppAndPipeline() {
  const app = new App();
  const pipeline = new CdkExpressPipeline();
  const wave1 = pipeline.addWave('global');
  const wave1stage1 = wave1.addStage('us-east-1');
  new ExpressStack(app, 'global', wave1stage1); // vpcs, certs, route53, iam

  const wave2 = pipeline.addWave('app');
  const wave2stage1 = wave2.addStage('us-east-1');
  new ExpressStack(app, 'api', wave2stage1);
  const wave2stage2 = wave2.addStage('eu-west-1');
  new ExpressStack(app, 'api', wave2stage2);

  return { app, pipeline };
}

describe('CDK Express Pipeline CI Configuration', () => {

  it('snapshot synth and build variants', async () => {
    const { pipeline } = getAppAndPipeline();

    function buildConfigVariant(buildConfig: BuildWorkflowPathConfig): GitHubWorkflowConfig {
      return {
        workingDir: '.',
        synth: {
          buildConfig: buildConfig,
          commands: [
            { dev: '' },
            { prod: '' },
          ],
        },
        diff: [],
        deploy: [],
      };
    }

    const tests: { name: string; config: GitHubWorkflowConfig }[] = [
      {
        name: 'preset-npm',
        config: buildConfigVariant({
          type: 'preset-npm',
        }),
      },
      {
        name: 'workflow',
        config: buildConfigVariant({
          type: 'workflow',
          workflow: {
            path: '.github/actions/build',
          },
        }),
      },
    ];

    for (const test of tests) {
      const resp = createGitHubWorkflows(test.config, pipeline.waves);
      const synthWorkflow = resp.find(w => w.fileName === 'actions/cdk-express-pipeline-synth/action.yml');
      expect(synthWorkflow).toMatchSnapshot(test.name);
    }
  });

  it('snapshot diff stack selector variants', async () => {
    const { pipeline } = getAppAndPipeline();

    function buildStackSelectorVariant(stackSelector: 'wave' | 'stage'): GitHubWorkflowConfig {
      return {
        workingDir: '.',
        synth: {
          buildConfig: {
            type: 'preset-npm',
          },
          commands: [
          ],
        },
        diff: [{
          on: {},
          stackSelector: stackSelector,
          writeAsComment: true,
          assumeRoleArn: '',
          assumeRegion: '',
          commands: [
            { dev: 'npm run cdk -- diff {stackSelector} --app=cdk.out/dev' },
          ],
        }],
        deploy: [],
      };
    }

    const tests: { name: string; config: GitHubWorkflowConfig }[] = [
      {
        name: 'wave',
        config: buildStackSelectorVariant('wave'),
      },
      {
        name: 'stage',
        config: buildStackSelectorVariant('stage'),
      },
    ];

    for (const test of tests) {
      const resp = createGitHubWorkflows(test.config, pipeline.waves);
      const synthWorkflow = resp.find(w => w.fileName === 'workflows/cdk-express-pipeline-diff.yml');
      expect((synthWorkflow?.content.json as any).jobs.diff__dev.strategy.matrix.include).toMatchSnapshot(test.name);
    }
  });

  it('snapshot all workflows for config', async () => {
    const app = new App();
    const pipeline = new CdkExpressPipeline();
    const wave1 = pipeline.addWave('global');
    const wave1stage1 = wave1.addStage('us-east-1');
    new ExpressStack(app, 'global', wave1stage1); // vpcs, certs, route53, iam

    const wave2 = pipeline.addWave('app');
    const wave2stage1 = wave2.addStage('us-east-1');
    new ExpressStack(app, 'api', wave2stage1);
    const wave2stage2 = wave2.addStage('eu-west-1');
    new ExpressStack(app, 'api', wave2stage2);

    const ghConfig: GitHubWorkflowConfig = {
      workingDir: '.',
      synth: {
        buildConfig: {
          type: 'preset-npm',
        },
        commands: [
          { dev: "npm run cdk -- synth '**' -c env=dev --output=cdk.out/dev" },
          { prod: "npm run cdk -- synth '**' -c env=prod --output=cdk.out/prod" },
        ],
      },
      diff: [{
        on: {
          pullRequest: {
            branches: ['main'],
          },
        },
        stackSelector: 'wave',
        writeAsComment: true,
        assumeRoleArn: 'arn:aws:iam::581184285249:role/githuboidc-git-hub-deploy-role',
        assumeRegion: 'us-east-1',
        commands: [
          { dev: 'npm run cdk -- diff {stackSelector} --app=cdk.out/dev' },
          { prod: 'npm run cdk -- diff {stackSelector} --app=cdk.out/prod' },
        ],
      }],
      deploy: [{
        id: 'dev',
        on: {
          pullRequest: {
            branches: ['main'],
          },
        },
        stackSelector: 'wave',
        assumeRoleArn: 'arn:aws:iam::581184285249:role/githuboidc-git-hub-deploy-role',
        assumeRegion: 'us-east-1',
        commands: [
          { dev: 'npm run cdk -- deploy {stackSelector} --app=cdk.out/dev --concurrency 10 --require-approval never --exclusively' },
        ],
      },
      {
        id: 'prod',
        on: {
          pullRequest: {
            branches: ['production'],
          },
        },
        stackSelector: 'wave',
        assumeRoleArn: 'arn:aws:iam::581184285249:role/githuboidc-git-hub-deploy-role',
        assumeRegion: 'us-east-1',
        commands: [
          { prod: 'npm run cdk -- deploy {stackSelector} --app=cdk.out/prod --concurrency 10 --require-approval never --exclusively' },
        ],
      }],
    };

    const resp = createGitHubWorkflows(ghConfig, pipeline.waves);
    for (const workflow of resp) {
      expect(workflow).toMatchSnapshot();
    }
  });


  it('generate workflows for demo-ts project', async () => {

    const app = new App();
    const expressPipeline = new CdkExpressPipeline();
    const wave1 = expressPipeline.addWave('Wave1');
    const wave1Stage1 = wave1.addStage('Stage1');
    const stackA = new ExpressStack(app, 'StackA', wave1Stage1);
    const stackB = new ExpressStack(app, 'StackB', wave1Stage1);
    stackB.addExpressDependency(stackA);
    const wave1Stage2 = wave1.addStage('Stage2');
    new ExpressStack(app, 'Stack2A', wave1Stage2);

    const wave2 = expressPipeline.addWave('Wave2');
    const wave2Stage1 = wave2.addStage('Stage1');
    new ExpressStack(app, 'StackC', wave2Stage1);

    const ghConfig: GitHubWorkflowConfig = {
      workingDir: '.',
      synth: {
        buildConfig: {
          type: 'preset-npm',
        },
        commands: [
          { dev: "npm run cdk -- synth '**' -c env=dev --output=cdk.out/dev" },
          { prod: "npm run cdk -- synth '**' -c env=prod --output=cdk.out/prod" },
        ],
      },
      diff: [{
        on: {
          pullRequest: {
            branches: ['main'],
          },
        },
        stackSelector: 'stage',
        writeAsComment: true,
        assumeRoleArn: 'arn:aws:iam::581184285249:role/githuboidc-git-hub-deploy-role',
        assumeRegion: 'us-east-1',
        commands: [
          { dev: 'npm run cdk -- diff {stackSelector} --app=cdk.out/dev' },
          { prod: 'npm run cdk -- diff {stackSelector} --app=cdk.out/prod' },
        ],
      }],
      deploy: [{
        id: 'dev',
        on: {
          pullRequest: {
            branches: ['main'],
          },
        },
        stackSelector: 'wave',
        assumeRoleArn: 'arn:aws:iam::581184285249:role/githuboidc-git-hub-deploy-role',
        assumeRegion: 'us-east-1',
        commands: [
          { dev: 'npm run cdk -- deploy {stackSelector} --app=cdk.out/dev --concurrency 10 --require-approval never --exclusively' },
        ],
      },
      {
        id: 'prod',
        on: {
          push: {
            branches: ['production'],
          },
        },
        stackSelector: 'wave',
        assumeRoleArn: 'arn:aws:iam::581184285249:role/githuboidc-git-hub-deploy-role',
        assumeRegion: 'us-east-1',
        commands: [
          { prod: 'npm run cdk -- deploy {stackSelector} --app=cdk.out/prod --concurrency 10 --require-approval never --exclusively' },
        ],
      }],
    };

    const resp = createGitHubWorkflows(ghConfig, expressPipeline.waves);
    for (const workflow of resp) {
      const filePath = path.join('/Users/rehanvandermerwe/Resend/Projects/cdk-express-pipeline-demo-ts/.github', workflow.fileName);
      const dirPath = path.dirname(filePath);
      if (!fsSync.existsSync(dirPath)) {
        fsSync.mkdirSync(dirPath, { recursive: true });
      }

      await fs.writeFile(filePath, stringify(workflow.content.json));
    }
  });
});