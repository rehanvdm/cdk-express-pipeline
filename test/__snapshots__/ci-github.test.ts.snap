// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CDK Express Pipeline CI Configuration snapshot all workflows for config 1`] = `
{
  "content": GithubWorkflow {
    "json": {
      "description": "Run CDK synth for a specific stack pattern and post results to PR",
      "inputs": {
        "cloud-assembly-directory": {
          "description": "The directory where the CDK cloud assembly should be synthesized to, must be the same as specified in the command",
          "required": true,
        },
        "command": {
          "description": "CDK synth command",
          "required": true,
        },
      },
      "name": "CDK Express Pipeline Synth Action",
      "runs": {
        "steps": [
          {
            "name": "Checkout repo",
            "uses": "actions/checkout@v4",
          },
          {
            "name": "Set up node",
            "uses": "actions/setup-node@v4",
            "with": {
              "cache": "npm",
              "node-version": 20,
            },
          },
          {
            "name": "Install dependencies",
            "run": "npm ci",
            "shell": "bash",
          },
          {
            "name": "CDK Synth",
            "run": "\${{ inputs.command }}",
            "shell": "bash",
          },
          {
            "name": "Cache Build",
            "uses": "actions/cache/save@v4",
            "with": {
              "key": "cache-build-\${{ github.sha }}-\${{ inputs.cloud-assembly-directory }}",
              "path": "\${{ inputs.cloud-assembly-directory }}/
node_modules/",
            },
          },
        ],
        "using": "composite",
      },
    },
  },
  "fileName": "actions/cdk-express-pipeline-synth/action.yml",
}
`;

exports[`CDK Express Pipeline CI Configuration snapshot all workflows for config 2`] = `
{
  "content": GithubWorkflow {
    "json": {
      "inputs": {
        "assume-region": {
          "description": "The AWS region to assume for the deploy operation",
          "required": true,
        },
        "assume-role-arn": {
          "description": "The ARN of the role to assume for the deploy operation",
          "required": true,
        },
        "cloud-assembly-directory": {
          "description": "The directory where the CDK cloud assembly is located",
          "required": true,
        },
        "command": {
          "description": "CDK diff command",
          "required": true,
        },
        "github-token": {
          "description": "GitHub token for posting comments",
          "required": true,
        },
        "job-name": {
          "description": "Name of the job, used to link to action/job logs in summaries",
          "required": false,
        },
        "stack-selector-patterns": {
          "description": "The value of the {stackSelector} replacement in the command",
          "required": true,
        },
      },
      "name": "CDK Diff Action",
      "runs": {
        "steps": [
          {
            "name": "Restore Build Cache",
            "uses": "actions/cache/restore@v4",
            "with": {
              "key": "cache-build-\${{ github.sha }}-\${{ inputs.cloud-assembly-directory }}",
              "path": "\${{ inputs.cloud-assembly-directory }}/
node_modules/",
            },
          },
          {
            "name": "Configure AWS credentials",
            "uses": "aws-actions/configure-aws-credentials@v4",
            "with": {
              "aws-region": "\${{ inputs.assume-region }}",
              "role-to-assume": "\${{ inputs.assume-role-arn }}",
            },
          },
          {
            "name": "CDK Diff Generate",
            "uses": "rehanvdm/cdk-express-pipeline-github-diff@v0.1.3",
            "with": {
              "cloud-assembly-directory": "\${{ inputs.cloud-assembly-directory }}",
              "github-token": "\${{ inputs.github-token }}",
              "job-name": "\${{ inputs.job-name }}",
              "mode": "generate",
              "stack-selectors": "\${{ inputs.stack-selector-patterns }}",
            },
          },
        ],
        "using": "composite",
      },
    },
  },
  "fileName": "actions/cdk-express-pipeline-diff/action.yml",
}
`;

exports[`CDK Express Pipeline CI Configuration snapshot all workflows for config 3`] = `
{
  "content": GithubWorkflow {
    "json": {
      "description": "Run CDK deploy for a specific stack pattern",
      "inputs": {
        "assume-region": {
          "description": "The AWS region to assume for the deploy operation",
          "required": true,
        },
        "assume-role-arn": {
          "description": "The ARN of the role to assume for the deploy operation",
          "required": true,
        },
        "cloud-assembly-directory": {
          "description": "The directory where the CDK cloud assembly is located",
          "required": true,
        },
        "command": {
          "description": "CDK deploy command",
          "required": true,
        },
        "stack-selector-patterns": {
          "description": "The value of the {stackSelector} replacement in the command",
          "required": true,
        },
      },
      "name": "CDK Deploy Action",
      "runs": {
        "steps": [
          {
            "name": "Restore Build Cache",
            "uses": "actions/cache/restore@v4",
            "with": {
              "key": "cache-build-\${{ github.sha }}-\${{ inputs.cloud-assembly-directory }}",
              "path": "\${{ inputs.cloud-assembly-directory }}/
node_modules/",
            },
          },
          {
            "name": "Configure AWS credentials",
            "uses": "aws-actions/configure-aws-credentials@v4",
            "with": {
              "aws-region": "\${{ inputs.assume-region }}",
              "role-to-assume": "\${{ inputs.assume-role-arn }}",
            },
          },
          {
            "name": "CDK Deploy Command",
            "run": "\${{ inputs.command }}",
            "shell": "bash",
          },
        ],
        "using": "composite",
      },
    },
  },
  "fileName": "actions/cdk-express-pipeline-deploy/action.yml",
}
`;

exports[`CDK Express Pipeline CI Configuration snapshot all workflows for config 4`] = `
{
  "content": GithubWorkflow {
    "json": {
      "jobs": {
        "diff__dev": {
          "name": "\${{ matrix.job-name }}",
          "needs": [
            "synth__dev",
          ],
          "permissions": {
            "actions": "write",
            "contents": "write",
            "id-token": "write",
            "pull-requests": "write",
          },
          "runs-on": "ubuntu-latest",
          "steps": [
            {
              "name": "Checkout repo",
              "uses": "actions/checkout@v4",
            },
            {
              "name": "Run CDK Express Pipeline Synth",
              "uses": "./.github/actions/cdk-express-pipeline-diff",
              "with": {
                "assume-region": "\${{ matrix.assume-region }}",
                "assume-role-arn": "\${{ matrix.assume-role-arn }}",
                "cloud-assembly-directory": "\${{ matrix.cloud-assembly-directory }}",
                "command": "\${{ matrix.command }}",
                "github-token": "\${{ secrets.GITHUB_TOKEN }}",
                "job-name": "\${{ matrix.job-name }}",
                "stack-selector-patterns": "\${{ matrix.stack-selector-patterns }}",
              },
            },
          ],
          "strategy": {
            "fail-fast": false,
            "matrix": {
              "include": [
                {
                  "assume-region": "us-east-1",
                  "assume-role-arn": "arn:aws:iam::581184285249:role/githuboidc-git-hub-deploy-role",
                  "cloud-assembly-directory": "cdk.out/dev",
                  "command": "npm run cdk -- diff global_* --app=cdk.out/dev",
                  "job-name": "Diff dev - global",
                  "stack-selector-patterns": "global_*",
                },
                {
                  "assume-region": "us-east-1",
                  "assume-role-arn": "arn:aws:iam::581184285249:role/githuboidc-git-hub-deploy-role",
                  "cloud-assembly-directory": "cdk.out/dev",
                  "command": "npm run cdk -- diff app_* --app=cdk.out/dev",
                  "job-name": "Diff dev - app",
                  "stack-selector-patterns": "app_*",
                },
              ],
            },
          },
        },
        "diff__prod": {
          "name": "\${{ matrix.job-name }}",
          "needs": [
            "synth__prod",
          ],
          "permissions": {
            "actions": "write",
            "contents": "write",
            "id-token": "write",
            "pull-requests": "write",
          },
          "runs-on": "ubuntu-latest",
          "steps": [
            {
              "name": "Checkout repo",
              "uses": "actions/checkout@v4",
            },
            {
              "name": "Run CDK Express Pipeline Synth",
              "uses": "./.github/actions/cdk-express-pipeline-diff",
              "with": {
                "assume-region": "\${{ matrix.assume-region }}",
                "assume-role-arn": "\${{ matrix.assume-role-arn }}",
                "cloud-assembly-directory": "\${{ matrix.cloud-assembly-directory }}",
                "command": "\${{ matrix.command }}",
                "github-token": "\${{ secrets.GITHUB_TOKEN }}",
                "job-name": "\${{ matrix.job-name }}",
                "stack-selector-patterns": "\${{ matrix.stack-selector-patterns }}",
              },
            },
          ],
          "strategy": {
            "fail-fast": false,
            "matrix": {
              "include": [
                {
                  "assume-region": "us-east-1",
                  "assume-role-arn": "arn:aws:iam::581184285249:role/githuboidc-git-hub-deploy-role",
                  "cloud-assembly-directory": "cdk.out/prod",
                  "command": "npm run cdk -- diff global_* --app=cdk.out/prod",
                  "job-name": "Diff prod - global",
                  "stack-selector-patterns": "global_*",
                },
                {
                  "assume-region": "us-east-1",
                  "assume-role-arn": "arn:aws:iam::581184285249:role/githuboidc-git-hub-deploy-role",
                  "cloud-assembly-directory": "cdk.out/prod",
                  "command": "npm run cdk -- diff app_* --app=cdk.out/prod",
                  "job-name": "Diff prod - app",
                  "stack-selector-patterns": "app_*",
                },
              ],
            },
          },
        },
        "print_diff": {
          "name": "Print Diff",
          "needs": [
            "diff__dev",
            "diff__prod",
          ],
          "permissions": {
            "actions": "write",
            "contents": "write",
            "id-token": "write",
            "pull-requests": "write",
          },
          "runs-on": "ubuntu-latest",
          "steps": [
            {
              "name": "Checkout repo",
              "uses": "actions/checkout@v4",
            },
            {
              "name": "Update PR Description",
              "uses": "rehanvdm/cdk-express-pipeline-github-diff@v0.1.3",
              "with": {
                "cloud-assemblies": "[{"header":"CDK Diff - dev","directory":"cdk.out/dev"},{"header":"CDK Diff - prod","directory":"cdk.out/prod"}]",
                "github-token": "\${{ secrets.GITHUB_TOKEN }}",
                "mode": "print",
              },
            },
          ],
        },
        "synth__dev": {
          "name": "Synth dev",
          "runs-on": "ubuntu-latest",
          "steps": [
            {
              "name": "Checkout repo",
              "uses": "actions/checkout@v4",
            },
            {
              "name": "Run CDK Express Pipeline Synth",
              "uses": "./.github/actions/cdk-express-pipeline-synth",
              "with": {
                "cloud-assembly-directory": "cdk.out/dev",
                "command": "npm run cdk -- synth '**' -c env=dev --output=cdk.out/dev",
              },
            },
          ],
        },
        "synth__prod": {
          "name": "Synth prod",
          "runs-on": "ubuntu-latest",
          "steps": [
            {
              "name": "Checkout repo",
              "uses": "actions/checkout@v4",
            },
            {
              "name": "Run CDK Express Pipeline Synth",
              "uses": "./.github/actions/cdk-express-pipeline-synth",
              "with": {
                "cloud-assembly-directory": "cdk.out/prod",
                "command": "npm run cdk -- synth '**' -c env=prod --output=cdk.out/prod",
              },
            },
          ],
        },
      },
      "name": "CDK Express Pipeline Diff - Dev",
      "on": {
        "pull_request": {
          "branches": [
            "main",
          ],
        },
      },
    },
  },
  "fileName": "workflows/cdk-express-pipeline-diff-dev.yml",
}
`;

exports[`CDK Express Pipeline CI Configuration snapshot all workflows for config 5`] = `
{
  "content": GithubWorkflow {
    "json": {
      "jobs": {
        "diff__prod": {
          "name": "\${{ matrix.job-name }}",
          "needs": [
            "synth__prod",
          ],
          "permissions": {
            "actions": "write",
            "contents": "write",
            "id-token": "write",
            "pull-requests": "write",
          },
          "runs-on": "ubuntu-latest",
          "steps": [
            {
              "name": "Checkout repo",
              "uses": "actions/checkout@v4",
            },
            {
              "name": "Run CDK Express Pipeline Synth",
              "uses": "./.github/actions/cdk-express-pipeline-diff",
              "with": {
                "assume-region": "\${{ matrix.assume-region }}",
                "assume-role-arn": "\${{ matrix.assume-role-arn }}",
                "cloud-assembly-directory": "\${{ matrix.cloud-assembly-directory }}",
                "command": "\${{ matrix.command }}",
                "github-token": "\${{ secrets.GITHUB_TOKEN }}",
                "job-name": "\${{ matrix.job-name }}",
                "stack-selector-patterns": "\${{ matrix.stack-selector-patterns }}",
              },
            },
          ],
          "strategy": {
            "fail-fast": false,
            "matrix": {
              "include": [
                {
                  "assume-region": "us-east-1",
                  "assume-role-arn": "arn:aws:iam::581184285249:role/githuboidc-git-hub-deploy-role",
                  "cloud-assembly-directory": "cdk.out/prod",
                  "command": "npm run cdk -- diff global_* --app=cdk.out/prod",
                  "job-name": "Diff prod - global",
                  "stack-selector-patterns": "global_*",
                },
                {
                  "assume-region": "us-east-1",
                  "assume-role-arn": "arn:aws:iam::581184285249:role/githuboidc-git-hub-deploy-role",
                  "cloud-assembly-directory": "cdk.out/prod",
                  "command": "npm run cdk -- diff app_* --app=cdk.out/prod",
                  "job-name": "Diff prod - app",
                  "stack-selector-patterns": "app_*",
                },
              ],
            },
          },
        },
        "print_diff": {
          "name": "Print Diff",
          "needs": [
            "diff__prod",
          ],
          "permissions": {
            "actions": "write",
            "contents": "write",
            "id-token": "write",
            "pull-requests": "write",
          },
          "runs-on": "ubuntu-latest",
          "steps": [
            {
              "name": "Checkout repo",
              "uses": "actions/checkout@v4",
            },
            {
              "name": "Update PR Description",
              "uses": "rehanvdm/cdk-express-pipeline-github-diff@v0.1.3",
              "with": {
                "cloud-assemblies": "[{"header":"CDK Diff - prod","directory":"cdk.out/prod"}]",
                "github-token": "\${{ secrets.GITHUB_TOKEN }}",
                "mode": "print",
              },
            },
          ],
        },
        "synth__prod": {
          "name": "Synth prod",
          "runs-on": "ubuntu-latest",
          "steps": [
            {
              "name": "Checkout repo",
              "uses": "actions/checkout@v4",
            },
            {
              "name": "Run CDK Express Pipeline Synth",
              "uses": "./.github/actions/cdk-express-pipeline-synth",
              "with": {
                "cloud-assembly-directory": "cdk.out/prod",
                "command": "npm run cdk -- synth '**' -c env=prod --output=cdk.out/prod",
              },
            },
          ],
        },
      },
      "name": "CDK Express Pipeline Diff - Prod",
      "on": {
        "pull_request": {
          "branches": [
            "production",
          ],
        },
      },
    },
  },
  "fileName": "workflows/cdk-express-pipeline-diff-prod.yml",
}
`;

exports[`CDK Express Pipeline CI Configuration snapshot all workflows for config 6`] = `
{
  "content": GithubWorkflow {
    "json": {
      "concurrency": {
        "cancel-in-progress": false,
        "group": "cdk-express-pipeline-deploy-dev",
      },
      "jobs": {
        "deploy__dev__app": {
          "name": "[dev] 🌊 app",
          "needs": [
            "deploy__dev__global",
          ],
          "permissions": {
            "actions": "write",
            "contents": "write",
            "id-token": "write",
          },
          "runs-on": "ubuntu-latest",
          "steps": [
            {
              "name": "Checkout repo",
              "uses": "actions/checkout@v4",
            },
            {
              "name": "Run CDK Express Pipeline Deploy",
              "uses": "./.github/actions/cdk-express-pipeline-deploy",
              "with": {
                "assume-region": "us-east-1",
                "assume-role-arn": "arn:aws:iam::581184285249:role/githuboidc-git-hub-deploy-role",
                "cloud-assembly-directory": "cdk.out/dev",
                "command": "npm run cdk -- deploy app_* --app=cdk.out/dev --concurrency 10 --require-approval never --exclusively",
                "stack-selector-patterns": "app_*",
              },
            },
          ],
        },
        "deploy__dev__global": {
          "name": "[dev] 🌊 global",
          "needs": [
            "synth__dev",
          ],
          "permissions": {
            "actions": "write",
            "contents": "write",
            "id-token": "write",
          },
          "runs-on": "ubuntu-latest",
          "steps": [
            {
              "name": "Checkout repo",
              "uses": "actions/checkout@v4",
            },
            {
              "name": "Run CDK Express Pipeline Deploy",
              "uses": "./.github/actions/cdk-express-pipeline-deploy",
              "with": {
                "assume-region": "us-east-1",
                "assume-role-arn": "arn:aws:iam::581184285249:role/githuboidc-git-hub-deploy-role",
                "cloud-assembly-directory": "cdk.out/dev",
                "command": "npm run cdk -- deploy global_* --app=cdk.out/dev --concurrency 10 --require-approval never --exclusively",
                "stack-selector-patterns": "global_*",
              },
            },
          ],
        },
        "synth__dev": {
          "name": "Synth dev",
          "runs-on": "ubuntu-latest",
          "steps": [
            {
              "name": "Checkout repo",
              "uses": "actions/checkout@v4",
            },
            {
              "name": "Run CDK Express Pipeline Synth",
              "uses": "./.github/actions/cdk-express-pipeline-synth",
              "with": {
                "cloud-assembly-directory": "cdk.out/dev",
                "command": "npm run cdk -- synth '**' -c env=dev --output=cdk.out/dev",
              },
            },
          ],
        },
      },
      "name": "CDK Express Pipeline Deploy - Dev",
      "on": {
        "pull_request": {
          "branches": [
            "main",
          ],
        },
      },
    },
  },
  "fileName": "workflows/cdk-express-pipeline-deploy-dev.yml",
}
`;

exports[`CDK Express Pipeline CI Configuration snapshot all workflows for config 7`] = `
{
  "content": GithubWorkflow {
    "json": {
      "concurrency": {
        "cancel-in-progress": false,
        "group": "cdk-express-pipeline-deploy-prod",
      },
      "jobs": {
        "deploy__prod__app": {
          "name": "[prod] 🌊 app",
          "needs": [
            "deploy__prod__global",
          ],
          "permissions": {
            "actions": "write",
            "contents": "write",
            "id-token": "write",
          },
          "runs-on": "ubuntu-latest",
          "steps": [
            {
              "name": "Checkout repo",
              "uses": "actions/checkout@v4",
            },
            {
              "name": "Run CDK Express Pipeline Deploy",
              "uses": "./.github/actions/cdk-express-pipeline-deploy",
              "with": {
                "assume-region": "us-east-1",
                "assume-role-arn": "arn:aws:iam::581184285249:role/githuboidc-git-hub-deploy-role",
                "cloud-assembly-directory": "cdk.out/prod",
                "command": "npm run cdk -- deploy app_* --app=cdk.out/prod --concurrency 10 --require-approval never --exclusively",
                "stack-selector-patterns": "app_*",
              },
            },
          ],
        },
        "deploy__prod__global": {
          "name": "[prod] 🌊 global",
          "needs": [
            "synth__prod",
          ],
          "permissions": {
            "actions": "write",
            "contents": "write",
            "id-token": "write",
          },
          "runs-on": "ubuntu-latest",
          "steps": [
            {
              "name": "Checkout repo",
              "uses": "actions/checkout@v4",
            },
            {
              "name": "Run CDK Express Pipeline Deploy",
              "uses": "./.github/actions/cdk-express-pipeline-deploy",
              "with": {
                "assume-region": "us-east-1",
                "assume-role-arn": "arn:aws:iam::581184285249:role/githuboidc-git-hub-deploy-role",
                "cloud-assembly-directory": "cdk.out/prod",
                "command": "npm run cdk -- deploy global_* --app=cdk.out/prod --concurrency 10 --require-approval never --exclusively",
                "stack-selector-patterns": "global_*",
              },
            },
          ],
        },
        "synth__prod": {
          "name": "Synth prod",
          "runs-on": "ubuntu-latest",
          "steps": [
            {
              "name": "Checkout repo",
              "uses": "actions/checkout@v4",
            },
            {
              "name": "Run CDK Express Pipeline Synth",
              "uses": "./.github/actions/cdk-express-pipeline-synth",
              "with": {
                "cloud-assembly-directory": "cdk.out/prod",
                "command": "npm run cdk -- synth '**' -c env=prod --output=cdk.out/prod",
              },
            },
          ],
        },
      },
      "name": "CDK Express Pipeline Deploy - Prod",
      "on": {
        "pull_request": {
          "branches": [
            "production",
          ],
        },
      },
    },
  },
  "fileName": "workflows/cdk-express-pipeline-deploy-prod.yml",
}
`;

exports[`CDK Express Pipeline CI Configuration snapshot diff stack selector variants: stage 1`] = `
[
  {
    "assume-region": "",
    "assume-role-arn": "",
    "cloud-assembly-directory": "cdk.out/dev",
    "command": "npm run cdk -- diff global_us-east-1_* --app=cdk.out/dev",
    "job-name": "Diff dev - global__us_east_1",
    "stack-selector-patterns": "global_us-east-1_*",
  },
  {
    "assume-region": "",
    "assume-role-arn": "",
    "cloud-assembly-directory": "cdk.out/dev",
    "command": "npm run cdk -- diff app_us-east-1_* --app=cdk.out/dev",
    "job-name": "Diff dev - app__us_east_1",
    "stack-selector-patterns": "app_us-east-1_*",
  },
  {
    "assume-region": "",
    "assume-role-arn": "",
    "cloud-assembly-directory": "cdk.out/dev",
    "command": "npm run cdk -- diff app_eu-west-1_* --app=cdk.out/dev",
    "job-name": "Diff dev - app__eu_west_1",
    "stack-selector-patterns": "app_eu-west-1_*",
  },
]
`;

exports[`CDK Express Pipeline CI Configuration snapshot diff stack selector variants: wave 1`] = `
[
  {
    "assume-region": "",
    "assume-role-arn": "",
    "cloud-assembly-directory": "cdk.out/dev",
    "command": "npm run cdk -- diff global_* --app=cdk.out/dev",
    "job-name": "Diff dev - global",
    "stack-selector-patterns": "global_*",
  },
  {
    "assume-region": "",
    "assume-role-arn": "",
    "cloud-assembly-directory": "cdk.out/dev",
    "command": "npm run cdk -- diff app_* --app=cdk.out/dev",
    "job-name": "Diff dev - app",
    "stack-selector-patterns": "app_*",
  },
]
`;
